### ISSUE TITLE
[Kata]SPDK vhost application failed to register memory region from virtual NVDIMM

### ISSUE THREAD

**Issue by xiaoxichen** [spdk/spdk#946]
Please use the issue tracker only for reporting suspected issues.

See [The SPDK Community Page](http://www.spdk.io/community/) for other SPDK communications channels.

SPDK vhost application failed to register memory region from virtual NVDIMM,  which is how Kata container provide image to the VM.
## Expected Behavior

Container should be able to use the vhost device from SPDK and memory registration should success
## Current Behavior

 am trying to access a malloc device in spdk vhost app
    through vhost-user-scsi-pci, in kata container, as a POC. Hugepage is
    configured and the chardev and scsi device was hardcoded as below:
       -chardev socket,id=char1,path=/var/tmp/vhost.0 -device
    vhost-user-scsi-pci,id=scsi2,chardev=char1
    
    From Kata side,
     though I can see the device inside container like
    
    root@6f4cf9b6f09a:/# lsblk --output
    "NAME,KNAME,MODEL,HCTL,SIZE,VENDOR,SUBSYSTEMS"
    NAME      KNAME   MODEL            HCTL        SIZE VENDOR   SUBSYSTEMS
    sda       sda     Malloc disk      1:0:0:0       1G INTEL
     block:scsi:virtio:pci
    
    
     and
    
    root@6f4cf9b6f09a:/sys/class/scsi_disk/1:0:0:0/device/block# ls
    sda
    
      But /dev/sda is not there.
    
    
    From SPDK side,   seems the NVDIMM which is used by Kata to hold the vm
    image , was passed to SPDK as a memory region,  however, the NVDIMM is not
    2MB aligned so spdk is complaining and Failed to register memory region
    
     vhost.c:1165:start_device: start_device with vid 0.
    
    spdk_vtophys_notify vaddr 0x7f553c00e238, len = 134217728
    paddr 0xffffffffffffffff
    SPDK_VTOPHYS_ERROR 0xffffffffffffffff
    paddr after pagemap:vaddr 0x7f5534000000-> paddr 0x2ea61ff000
    invalid paddr 0x2ea61ff000 - must be 2MB aligned
    
    
    I tried several things but without any luck,   Kata community do not have a
    way to switch away from NVDIMM and they love the DAX feature, then I cannot
    find a proper way to move the nvdimm to Hugepage to make it 2MB aligned...
   

## Steps to Reproduce

 The full qemu command :
    
    /usr/bin/qemu-vanilla-system-x86_64 -name
    sandbox-f57790e492edd4e9d081f9ccdcab9ccf7c923bd6438669f4ec4186bb7194e91a
    -uuid 34f304f2-feff-4d9c-84d0-9862a0a50619 -machine
    pc,accel=kvm,kernel_irqchip,nvdimm -cpu host -qmp
    unix:/run/vc/vm/f57790e492edd4e9d081f9ccdcab9ccf7c923bd6438669f4ec4186bb7194e91a/qmp.sock,server,nowait
    -m 2048M,slots=10,maxmem=386554M -device
    pci-bridge,bus=pci.0,id=pci-bridge-0,chassis_nr=1,shpc=on,addr=2,romfile=
    -device virtio-serial-pci,disable-modern=false,id=serial0,romfile= -device
    virtconsole,chardev=charconsole0,id=console0 -chardev
    socket,id=charconsole0,path=/run/vc/vm/f57790e492edd4e9d081f9ccdcab9ccf7c923bd6438669f4ec4186bb7194e91a/console.sock,server,nowait
    *-device nvdimm,id=nv0,memdev=mem0 -object
    memory-backend-file,id=mem0,mem-path=/usr/share/kata-containers/kata-containers-image_clearlinux_1.9.0-alpha0_agent_ffd19914b0.img,size=134217728*
    -device virtio-scsi-pci,id=scsi0,disable-modern=false,romfile= -object
    rng-random,id=rng0,filename=/dev/urandom -device
    virtio-rng,rng=rng0,romfile= -device
    virtserialport,chardev=charch0,id=channel0,name=agent.channel.0 -chardev
    socket,id=charch0,path=/run/vc/vm/f57790e492edd4e9d081f9ccdcab9ccf7c923bd6438669f4ec4186bb7194e91a/kata.sock,server,nowait
    -device
    virtio-9p-pci,disable-modern=false,fsdev=extra-9p-kataShared,mount_tag=kataShared,romfile=
    -fsdev
    local,id=extra-9p-kataShared,path=/run/kata-containers/shared/sandboxes/f57790e492edd4e9d081f9ccdcab9ccf7c923bd6438669f4ec4186bb7194e91a,security_model=none
    -netdev tap,id=network-0,vhost=on,vhostfds=3,fds=4 -device
    driver=virtio-net-pci,netdev=network-0,mac=02:42:ac:11:00:03,disable-modern=false,mq=on,vectors=4,romfile=
    -global kvm-pit.lost_tick_policy=discard -vga none -no-user-config
    -nodefaults -nographic -daemonize *-object
    memory-backend-file,id=dimm1,size=2048M,mem-path=/dev/hugepages,share=on,prealloc=on
    -numa node,memdev=dimm1* -kernel
    /usr/share/kata-containers/vmlinuz-4.19.52.45-35.container -append
    tsc=reliable no_timer_check rcupdate.rcu_expedited=1 i8042.direct=1
    i8042.dumbkbd=1 i8042.nopnp=1 i8042.noaux=1 noreplace-smp reboot=k
    console=hvc0 console=hvc1 iommu=off cryptomgr.notests net.ifnames=0
    pci=lastbus=0 root=/dev/pmem0p1
    rootflags=dax,data=ordered,errors=remount-ro ro rootfstype=ext4 quiet
    systemd.show_status=false panic=1 nr_cpus=48 agent.use_vsock=false
    init=/usr/lib/systemd/systemd systemd.unit=kata-containers.target
    systemd.mask=systemd-networkd.service systemd.mask=systemd-networkd.socket
    -pidfile
    /run/vc/vm/f57790e492edd4e9d081f9ccdcab9ccf7c923bd6438669f4ec4186bb7194e91a/pid
    -smp 1,cores=1,threads=1,sockets=48,maxcpus=48 *-chardev
    socket,id=char1,path=/var/tmp/vhost.0 -device
    vhost-user-scsi-pci,id=scsi2,chardev=char1*
## Context (Environment including OS version, SPDK version, etc.)

spdk: 1f133d72795f58d39f44eaff5eb601581abf2a20
 kata-runtime :1.9.0-alpha0

qemu-vanilla-system-x86_64 --version
QEMU emulator version 4.0.0
Copyright (c) 2003-2019 Fabrice Bellard and the QEMU Project developers

**Comment by xiaoxichen** [comment]
Related with https://github.com/spdk/spdk/issues/930

**Comment by changpe1** [comment]
Yeah, it's indeed an issue for SPDK vhost, although we already wrote the limitation in the documentation. We will take a look for this issue soon.

**Comment by peluse** [comment]
@changpe1 any update on this? The ebay team seems to be blocked on a pretty big project...

**Comment by changpe1** [comment]
@xiaoxichen Can you try with 'align=2M' parameter for your file based memory backend device? I think SPDK will not report unaligned memory issue, but I doubt SPDK can work with Kata or not, because the memory was provided by QEMU.

**Comment by xiaoxichen** [comment]
That parameters is generated by Kata runtime so basically we need to change kata

**Comment by xiaoxichen** [comment]
The QEMU memory is fine as we can specified below as we did for regular QEMU VM.

-object
memory-backend-file,id=dimm1,size=2048M,mem-path=/dev/hugepages,share=on,prealloc=on
-numa node,memdev=dimm1


Though by doing this the whole container footprint is in Hugemem as well,  which implies the whole(most of) the host memory need to be  configured as HugePage.

**Comment by changpe1** [comment]
That's right, if your Sandbox is started with hugetlbfs, so it can work SPDK, we already tested this before, and that's the recommended when using SDPK vhost target.

**Comment by peluse** [comment]
I figured I try and get kata setup here and start messing with it. The documentation is sorta all over the map but I believe I have docker configured to use kata-runtime correctly. @xiaoxichen can you add some more detail to your steps to reproduce, you mention that the qemu option suggestion about alignment was generated by qmeu. Can you elaborate on that? For example where did you find it, in a log or something? Earlier you commented "chardev and scsi device was hardcoded", where did you hardcode that? How are you starting your container from the cmd line? Sorry for all the questions :)

**Comment by xiaoxichen** [comment]
@peluse , KATA still do not have built-in  support  for vhost , this is how i tested it.

There is a kata configuration file (default to /usr/share/defaults/kata-containers/configuration.toml) where has
`path = "/usr/local/kata/opt/kata/bin/qemu-system-x86_64"`
so I put a script to wrap the real qemu, by adding the hardcoded chardev and scsi as parameters to the qemu

The configuration of memdev can also be hardcoded, or by setting the MemoryPath and HugePages
(https://github.com/kata-containers/runtime/blob/master/virtcontainers/persist/api/config.go)

**Comment by peluse** [comment]
@xiaoxichen are you by any chance going to the developer meetup next week?

**Comment by xiaoxichen** [comment]
@peluse  I wish I can but I based in Shanghai L(

**Comment by peluse** [comment]
@xiaoxichen starting to look at this again.  First, I guess I should ask if you've made any progress? Next, would it be possible to get more of a step-by-step type set of directions for the edits you made and exactly how to insert the go script wherever you are using it?

FYI I don't have the same path content that you have in your config file, here's what I think is the relevant section:
[hypervisor.qemu]
path = "/usr/bin/qemu-vanilla-system-x86_64"
kernel = "/usr/share/kata-containers/vmlinuz.container"
image = "/usr/share/kata-containers/kata-containers.img"
machine_type = "pc"

Can you comment?

**Comment by peluse** [comment]
@xiaoxichen or if you can provide the scripts you used to wrap qemu and add those parms that would be great too. I tried adding them in the path ="..." line in the .toml file but it doesn't like like that is going to work, I also just changed that path to a bash script that calls qemu with the parms and that doesn't work either.

**Comment by peluse** [comment]
@xiaoxichen ping... I've tried wrapping the qemu call w/a bash script and got some limited help on slack from the kata folks, still no luck. I have also had no luck finding the full qemu log for a failed attempt to wrap or the full cmd line used when things work correctly. You seem to have both of these figured out - please share :)

**Comment by xiaoxichen** [comment]
Hi @peluse , sorry for the late response. It is bad that my env was destroyed by someone accidentally but I tried to recall most from my log
1. create the script qemu.sh as a wrapper, which will append the vhost device parameters before other parameters from KATA. 
the configuration has two part, first part (before $@) is regarding the HugePage, and the second part is the char dev for vhost.

If you dont want to hardcode the first part, you can skip the first part but set https://github.com/kata-containers/runtime/blob/master/virtcontainers/persist/api/config.go#L75 to `/dev/hugepages` and https://github.com/kata-containers/runtime/blob/master/virtcontainers/persist/api/config.go#L125 to `True`

```
/usr/local/kata/opt/kata/bin/qemu-system-x86_64 -object memory-backend-file,id=mem,size=2G,mem-path=/dev/hugepages,share=on
-numa node,memdev=mem  $@ -chardev socket,id=char0,path=/var/tmp/vhost.0 -device vhost-user-scsi-pci,id=scsi0,chardev=char0 
```
2. Change the configuration file to use the wrapper script.
3. start a kata container, note that the memory size of the container must be exactly the same sa what specified in the script

**Comment by peluse** [comment]
@xiaoxichen thanks. It seems that no matter what I put in my script I get the same error:

docker: Error response from daemon: OCI runtime create failed: fail to launch qemu: exit status 1, error messages from qemu log: qemu-vanilla-system-x86_64: -object memory-backend-file,id=mem,size=2G,mem-path=/dev/hugepages,share=on
-numa node,memdev=mem  -name: invalid option: unknown.

and it doesn't matter if I use "stock qemu" or the vanilla one or whether the options I use are before or after the parm list, nothing seems to work.  Here's my script:

#!/bin/bash
/usr/bin/qemu-vanilla-system-x86_64 "-object memory-backend-file,id=mem,size=2G,mem-path=/dev/hugepages,share=on
-numa node,memdev=mem  $@ -chardev socket,id=char0,path=/var/tmp/vhost.0 -device vhost-user-scsi-pci,id=scsi0,chardev=char0"

and the kata folks don't seem to be of much help, I will try and bug them again tomorrow if I still don't make any progress.  @dong-liuliu can you try this on your end? Maybe it's just simple user error :)

**Comment by xiaoxichen** [comment]
@peluse maybe try to remove "-object memory-backend-file,id=mem,size=2G,mem-path=/dev/hugepages,share=on -numa node,memdev=mem -name"  from the script, but set 
```
MemoryPath = /dev/hugepages
HugePages = True
```
in the kata config files?

**Comment by peluse** [comment]
@xiaoxichen so optins before *or* after $@ results in that same error (invalid parm).

**Comment by peluse** [comment]
@xiaoxichen yes.  FYI I'm on Ubuntu 18.04 and installed kata/docker via these directions: https://github.com/kata-containers/documentation/blob/master/install/ubuntu-installation-guide.md

**Comment by peluse** [comment]
@xiaoxichen what is your env (distro, kernel version)? I think I should try a fresh install using your env

**Comment by peluse** [comment]
@xiaoxichen @dong-liuliu so I made a little progress, went to a whole different machine and got better results, I no longer get the invalid parm problem so must be something in my env on that system.  However, in just trying to add the hugepages I get this:

peluse@pels3:~$ sudo docker run hello-world
docker: Error response from daemon: OCI runtime create failed: failed to launch qemu: exit status 1, error messages from qemu log: qemu-vanilla-system-x86_64: total memory for NUMA nodes (0x100000000) should equal RAM size (0x80000000): unknown.

Note that in my .sh file and in my config file I specify 2049:
SCRIPT:
#!/bin/bash
/usr/bin/qemu-vanilla-system-x86_64 -object memory-backend-file,id=mem1,size=2048M,mem-path=/dev/hugepages,share=on,prealloc=on -numa node,memdev=mem1 "$@"

KATA CONFIG:
# Default memory size in MiB for SB/VM.
# If unspecified then it will be set 2048 MiB.
default_memory = 2048

And it doesn't matter what I change, it seems that it's always adding the two values to come up with "Total NUMA memory". So, for example, if I double the 2048 in the config file, it looks like this:

peluse@pels3:~$ sudo docker run hello-world
docker: Error response from daemon: OCI runtime create failed: failed to launch qemu: exit status 1, error messages from qemu log: qemu-vanilla-system-x86_64: total memory for NUMA nodes (0x180000000) should equal RAM size (0x100000000): unknown.
ERRO[0002] error waiting for container: context canceled

Any ideas?

**Comment by peluse** [comment]
@xiaoxichen @dong-liuliu more progress, I think. I could not get beyond that error above to hardocded the hugepages info as mentioned however @xiaoxichen you pointed me to the struct definitions, I modified where I thought the values were used in

	modified:   virtcontainers/factory/template/template.go
	modified:   virtcontainers/qemu.go

and it seemed to at least start w/o error.  Then I added the vhost info, same as above and what I start just using simple docker hello-world I do indeed get the SPDK output you mention (see below) however my docker app fails to start where it looks like you were able to get in and poke aro;und.  What image are you using?

vhost_session_mem_register: *INFO*: Registering VM memory for vtophys translation - 0x7fffd8000000 len:0x8000000
invalid paddr 0x73f9fd000 - must be 2MB aligned

(my docker error when I get the above SPDK error):

peluse@pels3:~/go/src/github.com/kata-containers/runtime$ sudo docker run hello-world
docker: Error response from daemon: OCI runtime create failed: Failed to check if grpc server is working: rpc error: code = Unavailable desc = transport is closing: unknown.
ERRO[0022] error waiting for container: context canceled

**Comment by xiaoxichen** [comment]
I use centos:latest

**Comment by peluse** [comment]
So, more progress. The hardcoded values I was using must have been incomplete.  When I make the following changes to the config file I can boot into an image with the only visible errors being the translation errors from SPDK. I'll look more into them shortly, I no longer see the alignment error message though, just a generic failure message.

enable_hugepages = true

# Enable file based guest memory support. The default is an empty string which
# will disable this feature. In the case of virtio-fs, this is enabled
# automatically and '/dev/shm' is used as the backing folder.
# This option will be ignored if VM templating is enabled.
file_mem_backend = "/dev/hugepages"

**Comment by peluse** [comment]
The first failure is in spdk_mem_map_translate() from:
          if (spdk_unlikely(vaddr & ~MASK_256TB))
returning map->default_translation which is 0.  Subsequent failures from from the next check in the same function:
         if (spdk_unlikely(!map_1gb))

I'll chat with others who know this code better than I do and see what they think...

**Comment by dong-liuliu** [comment]
Hi， @peluse 
As I can see, the issues for Kata and SPDK for Pmem are  *memory alignment  and *memory translation.
In most cases,  the pmem mapped in Kata are memory aligned. So  error "invalid paddr 0x73f9fd000 - must be 2MB aligned" occurs sometimes. And we can add alignment parameter to it as Changpeng suggested.
After Kata is booted, Kata can work well with SPDK. But only has some risk if Pmem part is used between Kata and SPDK vhost because of the memory translation issue.  Since that memory part seems rarely be used between Kata and SPDK vhost app in container, so it should be OK for users' unimportant workload.

I'm drafting patches for Kata runtime, then we can use docker to assign spdk vhost-user devices to kata container in an proper way.

**Comment by peluse** [comment]
@xiaoxichen @dong-liuliu Thanks! It seems your observations are consistent with mine and Xioxichen's with the exception of "After Kata is booted, Kata can work well with SPDK".  Neither of us were ale to see the SPDK vhost target after booting.  Can you expand on your comment?

**Comment by xiaoxichen** [comment]
I can see the SPDK Vhost target in `lsblk` however not in "/dev/sdX" and cannot do any IO. .

**Comment by dong-liuliu** [comment]
@peluse @xiaoxichen  A qemu wrapper with SPDK vhost-user target can only attach device to sandbox,  not container. And It requires cooperation between several Kata components to get vhost-user-blk device work in contianer. So both you and Xiaoxi only got Kata booted, but didn't see the device.
It seems the related code in Kata is incomplete for that requirement. so what I've tried to do is to create a proper way by code change to attach SPDK vhost-user-blk/scsi to Kata directly through docker cmdline.

**Comment by dong-liuliu** [comment]
Hi, Xiaoxi and Paul

A patch set to enable SPDK vhost-user devices to Kata container has been drafted.
[https://github.com/kata-containers/runtime/pull/2390](url)

I'm now drafting a document patch to Kata's domumentation repo (https://github.com/kata-containers/documentation). There the steps to run SPDK vhost-user device with Kata will be described.

**Comment by peluse** [comment]
Excellent, I'll look this week! Thanks @dong-liuliu

**Comment by dong-liuliu** [comment]
SPDK vhost-user-blk is already enabled in Kata [1.11.0-alpha1](https://github.com/kata-containers/runtime/releases/tag/1.11.0-alpha1) and will be included in upcoming Kata 1.11.0 release.
For more details, please visit [using-SPDK-vhostuser-and-kata](https://github.com/kata-containers/documentation/blob/master/use-cases/using-SPDK-vhostuser-and-kata.md)
